File: src/molgr/fallback/utils/no_metals/recovery.py
"""Explicit fallback recovery tiers for no-metal reconstruction."""

from __future__ import annotations

from typing import Callable, Iterable, cast

from openbabel import openbabel as ob
from openbabel import pybel

from molgr.fallback.stages.break_bond import break_deformed_ene, break_one_bond
from molgr.fallback.state import OmolStateMachine, ReconstructionState
from molgr.fallback.utils import resonance as resonance_utils


def _clone_omol(omol: pybel.Molecule) -> pybel.Molecule:
    return pybel.Molecule(ob.OBMol(cast(ob.OBMol, omol.OBMol)))


def _deduplicate(states: Iterable[ReconstructionState]) -> list[ReconstructionState]:
    unique: dict[tuple[object, int], ReconstructionState] = {}
    for state in states:
        key = resonance_utils.build_resonance_state_key(state.omol), state.given_charge
        unique.setdefault(key, state)
    return list(unique.values())


def enumerate_deformed_pi_recovery_seeds(
    states: Iterable[ReconstructionState],
    *,
    break_stage: Callable[..., tuple[pybel.Molecule, bool]] | None = None,
) -> list[ReconstructionState]:
    """Lower deformed bonds whose stage explicitly updates endpoint electrons."""

    break_stage = break_deformed_ene if break_stage is None else break_stage
    recovered: list[ReconstructionState] = []
    for state in states:
        machine = OmolStateMachine.from_reconstruction_state(state).branch(
            None,
            omol=_clone_omol(state.omol),
        )
        hit = machine.run_omol_stage(
            "recover_deformed_pi_bonds",
            break_stage,
            machine.given_charge,
            state.total_radical_electrons,
            5.0,
        )
        if not hit:
            continue
        machine.annotate(None, recovery_tier=1, recovery_strategy="deformed_pi_bonds")
        recovered.append(machine.freeze_like(state))
    return _deduplicate(recovered)


def enumerate_bond_break_recovery_seeds(
    states: Iterable[ReconstructionState],
    *,
    break_stage: Callable[..., tuple[pybel.Molecule, int, bool]] | None = None,
) -> list[ReconstructionState]:
    """Apply bond-breaking rules whose stages explicitly update endpoint electrons."""

    break_stage = break_one_bond if break_stage is None else break_stage
    recovered: list[ReconstructionState] = []
    for state in states:
        machine = OmolStateMachine.from_reconstruction_state(state).branch(
            None,
            omol=_clone_omol(state.omol),
        )
        hit = machine.run_omol_charge_stage(
            "recover_by_breaking_bonds",
            break_stage,
            state.total_radical_electrons,
        )
        if not hit:
            continue
        machine.annotate(None, recovery_tier=2, recovery_strategy="bond_break")
        recovered.append(machine.freeze_like(state))
    return _deduplicate(recovered)


__all__ = [
    "enumerate_bond_break_recovery_seeds",
    "enumerate_deformed_pi_recovery_seeds",
]

File: src/cpp/src/utils/no_metals/recovery.cpp
#include "molgr/utils/no_metals/recovery.h"

#include "molgr/stages/break_bond.h"
#include "molgr/stages/fresh.h"
#include "molgr/utils/conversions.h"
#include "molgr/utils/resonance.h"

#include <memory>
#include <set>
#include <utility>
#include <vector>

namespace
{
    using StateKey = std::pair<molgr::resonance::ResonanceStateKey, int>;

    std::vector<molgr::state::ReconstructionState> Deduplicate(
        std::vector<molgr::state::ReconstructionState> states)
    {
        std::set<StateKey> seen;
        std::vector<molgr::state::ReconstructionState> unique;
        for (auto &state : states)
        {
            StateKey key{molgr::resonance::BuildResonanceStateKey(state.Mol()), state.given_charge};
            if (seen.insert(std::move(key)).second)
            {
                unique.push_back(std::move(state));
            }
        }
        return unique;
    }
}

namespace molgr
{
    namespace no_metals
    {
        namespace recovery
        {
            std::vector<molgr::state::ReconstructionState> EnumerateDeformedPiRecoverySeeds(
                const std::vector<molgr::state::ReconstructionState> &states)
            {
                std::vector<molgr::state::ReconstructionState> recovered;
                for (const auto &state : states)
                {
                    auto machine = molgr::state::OmolStateMachine::FromReconstructionState(state);
                    auto next_mol = std::make_shared<OpenBabel::OBMol>(
                        molgr::utils::CloneMolTopologyOnly(state.Mol()));
                    machine = machine.Branch(std::nullopt, std::move(next_mol));
                    const bool hit = machine.RunOmolStage(
                        "recover_deformed_pi_bonds",
                        reconstruct::BreakDeformedEne,
                        machine.given_charge,
                        state.total_radical_electrons,
                        5.0);
                    if (!hit)
                    {
                        continue;
                    }
                    machine.RunOmolStage(
                        "refresh_electronic_labels_after_recovery",
                        reconstruct::FreshOmolChargeRadical);
                    machine.metadata["recovery_tier"] = 1;
                    machine.metadata["recovery_strategy"] = std::string("deformed_pi_bonds");
                    recovered.push_back(machine.FreezeLike(state));
                }
                return Deduplicate(std::move(recovered));
            }

            std::vector<molgr::state::ReconstructionState> EnumerateBondBreakRecoverySeeds(
                const std::vector<molgr::state::ReconstructionState> &states)
            {
                std::vector<molgr::state::ReconstructionState> recovered;
                for (const auto &state : states)
                {
                    auto machine = molgr::state::OmolStateMachine::FromReconstructionState(state);
                    auto next_mol = std::make_shared<OpenBabel::OBMol>(
                        molgr::utils::CloneMolTopologyOnly(state.Mol()));
                    machine = machine.Branch(std::nullopt, std::move(next_mol));
                    const bool hit = machine.RunOmolChargeStage(
                        "recover_by_breaking_bonds",
                        reconstruct::BreakOneBond,
                        state.total_radical_electrons);
                    if (!hit)
                    {
                        continue;
                    }
                    machine.RunOmolStage(
                        "refresh_electronic_labels_after_recovery",
                        reconstruct::FreshOmolChargeRadical);
                    machine.metadata["recovery_tier"] = 2;
                    machine.metadata["recovery_strategy"] = std::string("bond_break");
                    recovered.push_back(machine.FreezeLike(state));
                }
                return Deduplicate(std::move(recovered));
            }
        }
    }
}
